nsecx.go

```diff
@@ -48,62 +48,49 @@ func HashName(label string, ha uint8, iter uint16, salt string) string {
 	return toBase32(nsec3)
 }
 
-// Denialer is an interface that should be implemented by types that are used to denial
-// answers in DNSSEC.
-type Denialer interface {
-	// Cover will check if the (unhashed) name is being covered by this NSEC or NSEC3.
-	Cover(name string) bool
-	// Match will check if the ownername matches the (unhashed) name for this NSEC3 or NSEC3.
-	Match(name string) bool
-}
-
-// Cover implements the Denialer interface.
-func (rr *NSEC) Cover(name string) bool {
-	return true
-}
-
-// Match implements the Denialer interface.
-func (rr *NSEC) Match(name string) bool {
-	return true
-}
-
-// Cover implements the Denialer interface.
+// Cover returns true if a name is covered by the NSEC3 record
 func (rr *NSEC3) Cover(name string) bool {
-	// FIXME(miek): check if the zones match
-	// FIXME(miek): check if we're not dealing with parent nsec3
-	hname := HashName(name, rr.Hash, rr.Iterations, rr.Salt)
-	labels := Split(rr.Hdr.Name)
+	nameHash := HashName(name, rr.Hash, rr.Iterations, rr.Salt)
+	owner := strings.ToUpper(rr.Hdr.Name)
+	labels := Split(owner)
 	if len(labels) < 2 {
 		return false
 	}
-	hash := strings.ToUpper(rr.Hdr.Name[labels[0] : labels[1]-1]) // -1 to remove the dot
-	if hash == rr.NextDomain {
-		return false // empty interval
-	}
-	if hash > rr.NextDomain { // last name, points to apex
-		// hname > hash
-		// hname > rr.NextDomain
-		// TODO(miek)
+	if !strings.HasSuffix(strings.ToUpper(name), owner[labels[1]:]) { // name is outside zone
```

Comment:
This checks that the provided name is a subdomain of the parent of the owner name of the record (i.e. is in the owners zone) which avoids weird cases where the hashed name may appear to be covered but isn't actually in the relevant zone. `dns.IsSubdomain` would work perfectly here except it seems to be case insensitive so it will fail to detect that `asd.test.com` is a subdomain of `TEST.com` which could introduce some hard to debug issues.

I'd be happy to make `IsSubdomain` insensitive if that works for you, it should just require two `strings.ToLower` operations, which are pretty cheap in the overall picture.